nautobot/dcim/factory.py

```diff
@@ -160,32 +166,49 @@ class Meta:
             "has_site",
             "has_tenant",
             "has_description",
+            "_parent",
         )
 
     name = factory.LazyAttributeSequence(lambda l, n: f"{l.location_type.name}-{n:02d}")
     status = random_instance(lambda: Status.objects.get_for_model(Location), allow_null=False)
 
-    has_parent = factory.LazyAttribute(lambda l: bool(l.location_type.parent))
-    parent = factory.Maybe(
-        "has_parent",
-        factory.LazyAttribute(
-            lambda l: factory.random.randgen.choice(Location.objects.filter(location_type=l.location_type.parent))
-            if Location.objects.count()
-            else None
-        ),
-    )
-
-    has_site = factory.LazyAttribute(lambda l: not bool(l.has_parent))
+    @factory.iterator
+    def location_type():  # pylint: disable=no-method-argument
+        lts = LocationType.objects.all()
+        for lt in lts:
+            yield lt
+
+    @factory.lazy_attribute
+    def has_parent(self):
+        if not self.location_type.nestable:
+            return bool(self.location_type.parent)
+        else:
+            return Faker().boolean()
+
+    @factory.lazy_attribute
+    def _parent(self):
+        if not self.location_type.nestable:
+            candidate_parents = Location.objects.filter(location_type=self.location_type.parent)
+        elif self.location_type.parent is not None:
+            candidate_parents = Location.objects.filter(
+                Q(location_type=self.location_type.parent) | Q(location_type=self.location_type)
+            )
+        else:
+            candidate_parents = Location.objects.filter(location_type=self.location_type)
+
+        if candidate_parents.exists():
+            return factory.random.randgen.choice(candidate_parents)
+        return None
+
+    parent = factory.Maybe("has_parent", _parent)
```

Comment:
These could be condensed to 1 `lazy_attribute` method

```suggestion
    @factory.lazy_attribute
    def parent(self):

        # 50% chance to have a parent
        if not Faker().pybool():
            return

        candidate_parents = Q()
        if self.location_type.parent is not None:
            candidate_parents |= Q(location_type=self.location_type.parent)
            if self.location_type.nestable:
                candidate_parents |= Q(location_type=self.location_type)
        else:
            candidate_parents |= Q(location_type=self.location_type)

        parents = Location.objects.filter(candidate_parents)
        if parents.exists():
            return factory.random.randgen.choice(parents)
        return None
```

Reply:
```
if not Faker().pybool():
            return
```
This will return None for some of the locations that def should have a parent. So I think we should check `self.location_type.parent` first and use that statement iff the location is nestable.

Reply:
Yeah I see now that a parent is required if `location_type.parent` is set. I see another problem with my suggestion where it's not accounting for nestable location types that don't have a parent. The possible permutations are:


| is nestable  |  has parent |
| -------------|------------- |
| False | False |
| True | False |
| False | True |
| True | True |


Maybe this?


```py
    @factory.lazy_attribute
    def parent(self):

        # 50% chance to not have a parent
        if not self.location_type.parent and Faker().pybool():
            return

        candidate_parents = Q()
        if self.location_type.parent:
            candidate_parents |= Q(location_type=self.location_type.parent)
        if self.location_type.nestable:
            candidate_parents |= Q(location_type=self.location_type)
        parents = Location.objects.filter(candidate_parents)
        if parents.exists():
            return factory.random.randgen.choice(parents)
        return None
```